File: packages/shared/src/problem-details.ts
import { z } from "zod";

const ProblemCodeSchema = z.string().regex(/^[a-z][a-z0-9_]*$/);

export const ApiFieldErrorSchema = z.strictObject({
  /** JSON Pointer to the invalid request member; an empty string addresses the document. */
  pointer: z.string().regex(/^(?:$|\/)/),
  code: ProblemCodeSchema,
  message: z.string().trim().min(1),
});
export type ApiFieldError = z.infer<typeof ApiFieldErrorSchema>;

/** RFC 9457 problem details plus stable machine codes and field errors. */
export const ApiProblemSchema = z.strictObject({
  type: z.string().trim().min(1).default("about:blank"),
  title: z.string().trim().min(1),
  status: z.number().int().min(400).max(599),
  detail: z.string().trim().min(1).optional(),
  instance: z.string().trim().min(1).optional(),
  code: ProblemCodeSchema,
  errors: z.array(ApiFieldErrorSchema).optional(),
});
export type ApiProblem = z.infer<typeof ApiProblemSchema>;

File: packages/shared/python/iopsych_contracts/problem_details.py
"""Standard API problem-detail response contracts."""

from typing import Annotated

from pydantic import BaseModel, ConfigDict, Field, StringConstraints

ProblemCode = Annotated[str, StringConstraints(pattern=r"^[a-z][a-z0-9_]*$")]


class ApiFieldError(BaseModel):
    """A request validation issue located with a JSON Pointer."""

    model_config = ConfigDict(extra="forbid")

    pointer: Annotated[str, StringConstraints(pattern=r"^(?:$|/)")]
    code: ProblemCode
    message: Annotated[str, StringConstraints(strip_whitespace=True, min_length=1)]


class ApiProblem(BaseModel):
    """RFC 9457 problem details plus stable machine-readable extensions."""

    model_config = ConfigDict(extra="forbid")

    type: Annotated[str, StringConstraints(strip_whitespace=True, min_length=1)] = "about:blank"
    title: Annotated[str, StringConstraints(strip_whitespace=True, min_length=1)]
    status: Annotated[int, Field(strict=True, ge=400, le=599)]
    detail: Annotated[str, StringConstraints(strip_whitespace=True, min_length=1)] | None = None
    instance: Annotated[str, StringConstraints(strip_whitespace=True, min_length=1)] | None = None
    code: ProblemCode
    errors: list[ApiFieldError] | None = None
